src/util/codec/number.rs

```diff
@@ -219,6 +222,112 @@ pub trait NumberDecoder: Read {
 
 impl<T: Read> NumberDecoder for T {}
 
+type Bytes<'a> = &'a [u8];
+
+const SIZE_OF_U64: usize = 8;
+const SIZE_OF_U32: usize = 4;
+const SIZE_OF_U16: usize = 2;
+const SIZE_OF_F64: usize = 8;
+
+macro_rules! read_num_bytes {
+    ($size:expr, $data:expr, $fn:path) => {{
+        if $data.len() < $size {
+            return Err(Error::Io(io::Error::new(ErrorKind::UnexpectedEof, "eof")));
+        }
+        let buf = &$data[0..$size];
+        *$data = &$data[$size..];
+        Ok($fn(buf))
+    }};
+}
+/// `decode_i64` decodes value encoded by `encode_i64` before.
+fn decode_i64(data: &mut Bytes) -> Result<i64> {
+    decode_u64(data).map(order_decode_i64)
+}
+
+/// `decode_i64_desc` decodes value encoded by `encode_i64_desc` before.
+fn decode_i64_desc(data: &mut Bytes) -> Result<i64> {
+    decode_u64_desc(data).map(order_decode_i64)
+}
+
+/// `decode_u64` decodes value encoded by `encode_u64` before.
+pub fn decode_u64(data: &mut Bytes) -> Result<u64> {
+    read_num_bytes!(SIZE_OF_U64, data, BigEndian::read_u64)
+}
+
+/// `decode_u64_desc` decodes value encoded by `encode_u64_desc` before.
+pub fn decode_u64_desc(data: &mut Bytes) -> Result<u64> {
+    let v = decode_u64(data)?;
+    Ok(!v)
+}
+
+/// `decode_var_i64` decodes value encoded by `encode_var_i64` before.
+fn decode_var_i64(data: &mut Bytes) -> Result<i64> {
+    let v = decode_var_u64(data)?;
+    let mut vx = v >> 1;
+    if v & 1 != 0 {
+        vx = !vx;
+    }
+    Ok(vx as i64)
+}
+
+/// `decode_var_u64` decodes value encoded by `encode_var_u64` before.
+pub fn decode_var_u64(data: &mut Bytes) -> Result<u64> {
+    let (mut x, mut s, mut i) = (0, 0, 0);
+    while i < data.len() {
```

Comment:
Generally the number is less than 516. So you can check if decode one byte is enough first.